File: 백준/Gold/24620. Sleeping in Class/Sleeping in Class.py
import sys
input = sys.stdin.readline

def f(n): #배열의 합의 약수 구하기
    num = []
    for i in range(1, int(n ** 0.5) + 1):
        if n % i == 0:
            num.append(i)
            if i * i != n:  # 제곱수 예외 처리
                num.append(n // i)
    return num

t = int(input())  # 1 ~ 10개
for ss in range(t):
    n = int(input())  # 수업기간 1~100000
    arr = list(map(int, input().split()))

    s = sum(arr) # 배열의 합
    num = f(s) # 합의 약수 배열
    num.sort()

    ans = 0
    for i in num:
        arrsum = 0
        tf = True

        for j in range(n):
            arrsum += arr[j]
            if arrsum > i:
                tf = False
                break

            elif arrsum == i:
                arrsum = 0

        if tf:
            ans = n - (s//i)
            break

    print(ans)

File: 백준/Gold/24620. Sleeping in Class/Sleeping in Class.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
    // 배열의 합의 약수 구하기
    public static ArrayList<Integer> getDivisors(int n) {
        ArrayList<Integer> divisors = new ArrayList<>();
        for (int i = 1; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                divisors.add(i);
                if (i * i != n) { // 제곱수 예외 처리
                    divisors.add(n / i);
                }
            }
        }
        return divisors;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int t = Integer.parseInt(br.readLine()); // 테스트 케이스의 수 (1 ~ 10)
        for (int ss = 0; ss < t; ss++) {
            int n = Integer.parseInt(br.readLine()); // 수업 기간 (1 ~ 100000)
            st = new StringTokenizer(br.readLine());
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }

            int s = 0; // 배열의 합
            for (int value : arr) {
                s += value;
            }
            ArrayList<Integer> divisors = getDivisors(s); // 합의 약수 배열
            Collections.sort(divisors);

            int ans = 0;
            for (int divisor : divisors) {
                int arrSum = 0;
                boolean tf = true;

                for (int j = 0; j < n; j++) {
                    arrSum += arr[j];
                    if (arrSum > divisor) {
                        tf = false;
                        break;
                    } else if (arrSum == divisor) {
                        arrSum = 0;
                    }
                }

                if (tf) {
                    ans = n - (s / divisor);
                    break;
                }
            }
            System.out.println(ans);
        }
    }
}
